src/build.ml

```diff
@@ -188,12 +199,13 @@ let get_prog = function
     >>> dyn_paths (arr (function Error _ -> [] | Ok x -> [x]))
 
 let prog_and_args ?(dir=Path.root) prog args =
-  Paths (Arg_spec.add_deps args Path.Set.empty)
+  let deps = of_deps (Arg_spec.static_deps args) in
```

Comment:
It looks like `of_deps` already registers dependencies. So you shouldn't need to register them a second time.

I pushed a version that should only register once.

Reply:
In fact, this is done as an optimization. We first extract the static dependencies but wait for static _and_ dynamic before running the action. Static dependencies are fairly useful as they give us a better parallelism in general.

Reply:
Sorry, I'm not following. What is done as an optimization? Your version optimization over mine? But I think mine just removes double-counting of dependencies. Let's return to this on Monday when I can think better.

Reply:
Your version is the one that optimizes correctly. `dyn_paths` and the new `dyn_deps` should be avoided.

> Let's return to this on Monday when I can think better.

Deal.

Reply:
Yeah, OK, I get it. :-)

Speaking of `of_deps`, maybe it should return a `unit` instead of collecting the paths, to make it more clear that it registers the dependencies rather than just compute them? What do you think?

Reply:
@aalekseyev done. Also, I've removed `of_deps` from the public interface